### PR TITLE
dell-dock: Fix some of the Coverity-reported issues

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/dell-dock/fu-dell-dock-plugin.c [hunk 1]
```diff
@@ -75,9 +75,57 @@ fu_dell_dock_plugin_get_dock_type(FuDevice *device, GError **error)
...
+				    guint8 dock_type,
+				    GError **error)
 {
 	FuContext *ctx = fu_plugin_get_context(plugin);
+	const gchar *instance_id;
+	g_autoptr(FuDellDockMst) mst_device = fu_dell_dock_mst_new(ctx);
+
+	instance_id = fu_dell_dock_get_instance_id(dock_type, dock_component_mst, 0, 0);
+	fu_device_add_instance_id(FU_DEVICE(mst_device), instance_id);
+	fu_device_add_child(FU_DEVICE(ec_device), FU_DEVICE(mst_device));
```
Comment (CragW) [inline plugins/dell-dock/fu-dell-dock-plugin.c:89]: Warning message is seen at add_child time if no pre-assigned guid:
```
FuEngine             no GUIDs for device VMM5331 in Dell dock [106be225d4ab81a272bee0ae24a9fab93a54edfe]
```

File: plugins/dell-dock/fu-dell-dock-plugin.c [hunk 2]
```diff
@@ -75,9 +75,57 @@ fu_dell_dock_plugin_get_dock_type(FuDevice *device, GError **error)
...
+{
+	FuContext *ctx = fu_plugin_get_context(plugin);
+	const gchar *instance_id;
+	guint16 dock_variant;
+	g_autoptr(FuDellDockStatus) status_device = fu_dell_dock_status_new(ctx);
+
+	dock_variant = fu_dell_dock_module_is_usb4(FU_DEVICE(ec_device)) ? 1 : 0;
+	instance_id = fu_dell_dock_get_instance_id(dock_type, dock_component_pkg, 0, dock_variant);
+	fu_device_add_instance_id(FU_DEVICE(status_device), instance_id);
+	fu_device_add_child(FU_DEVICE(ec_device), FU_DEVICE(status_device));
```
Comment (CragW) [inline plugins/dell-dock/fu-dell-dock-plugin.c:109]: same here.
```
FuEngine             no GUIDs for device Package level of Dell dock [106be225d4ab81a272bee0ae24a9fab93a54edfe]
```

File: plugins/dell-dock/fu-dell-dock-plugin.c [hunk 3]
```diff
@@ -75,9 +75,57 @@ fu_dell_dock_plugin_get_dock_type(FuDevice *device, GError **error)
...
+static gboolean
+fu_dell_dock_plugin_probe_ec_v1_tbt3(FuPlugin *plugin, FuDevice *ec_device, GError **error)
+{
+	g_autoptr(FuDellDockTbt) tbt_device = NULL;
+
+	if (!fu_dell_dock_ec_needs_tbt(FU_DEVICE(ec_device)))
+		return TRUE;
+	tbt_device = fu_dell_dock_tbt_new(fu_device_get_proxy(ec_device));
+	fu_device_add_instance_id(FU_DEVICE(tbt_device), DELL_DOCK_TBT3);
+	fu_device_add_child(FU_DEVICE(ec_device), FU_DEVICE(tbt_device));
```
Comment (CragW) [inline plugins/dell-dock/fu-dell-dock-plugin.c:122]: here too.

**Comment by CragW** [comment]
Thanks for fixing the issues.

File: plugins/dell-dock/fu-dell-dock-plugin.c (hunk 1 above)
Comment (hughsie) [reply plugins/dell-dock/fu-dell-dock-plugin.c:89]: If you do fu_device_probe() before adding the child does that fix things? The lifecycle of the dock components is very hard to understand.

Comment (CragW) [reply plugins/dell-dock/fu-dell-dock-plugin.c:89]: adding `guid` manually is the old way it used to mitigate this issue, which is missing in this PR. i.e.: https://github.com/fwupd/fwupd/blob/main/plugins/dell-dock/fu-dell-dock-plugin.c#L104.

in upcoming k2 I put the relationship in device->setup:
https://github.com/fwupd/fwupd/pull/7355/files#diff-11355388afc091f3fc67e4b4be9bf11e3aacd8fcc5c5a812ca767e1e463ce236R66

File: plugins/dell-dock/fu-dell-dock-plugin.c [hunk 4]
```diff
@@ -51,101 +57,104 @@ struct _FuDellDockPlugin {
...
 {
+	FuContext *ctx = fu_plugin_get_context(plugin);
+	const gchar *instance_id;
+	g_autoptr(FuDellDockMst) mst_device = fu_dell_dock_mst_new(ctx);
 	g_autoptr(FuDeviceLocker) locker = NULL;
 
-	locker = fu_device_locker_new(device, error);
+	instance_id = fu_dell_dock_get_instance_id(dock_type, dock_component_mst, 0, 0);
+	fu_device_add_instance_id(FU_DEVICE(mst_device), instance_id);
+	locker = fu_device_locker_new(FU_DEVICE(mst_device), error);
```
Comment (CragW) [inline plugins/dell-dock/fu-dell-dock-plugin.c:82]: after some tests, the `mst` device in the `setup` lifecycle time looks for its `parent` and if `parent` is unset at that time will end up in a segmentation fault.

**Comment by hughsie** [comment]
@CragW I've reverted 98be21bf2268eaa44b55ce1f5d48767032189045 and ca877b42ecb73ff8a4e102711dc5e9359c17695f -- coverity is really unhappy and I'm quite displeased at the added complexity and the risk of breaking things. I think the new dell-dock2 plugin for the atomic dock makes those commits unnecessary now. If we want to bring some of that functionality back to dell-dock we need a lot more documentation (in README.md or in the code itself) -- in general the code should look in style something like the hughsie/coverity-dell-dock branch with the namespaced enums and the use of the high level primitives -- and also be valgrind clean -- i.e. not leaking memory.

**Pull request closed without merging**